src/ext/detective.lisp

```diff
@@ -0,0 +1,282 @@
+  (setf (buffer-value buffer 'references) value))
+
+(defgeneric search-references (search-class))
+
+(defmethod search-references ((search search-regex))
+  (labels ((find-ref (regex class)
+             (with-point ((p (buffer-start-point (point-buffer (current-point)))))
+               (loop :for position = (search-forward-regexp p (capture-regex-regex regex))
+                     :while position
+                     :collect (funcall (capture-regex-function regex) 
+                                       (copy-point position)
+                                       class)))))
+  (with-slots (function-regex
+               package-regex
+               class-regex
+               variable-regex
+               misc-regex)
+      search
+    (let ((slots 
+            (list (cons (cons "functions" function-regex):function-reference)
+                  (cons (cons "classes" class-regex) :class-reference)
+                  (cons (cons "packages" package-regex) :package-reference)
+                  (cons (cons "variables" variable-regex) :variable-reference)
+                  (cons (cons "misc" misc-regex) :misc-reference))))
+    
+    (setf (buffer-references (current-buffer))
+          (make-hash-table :test 'equal))
+
+      (loop :for ((id . regex) . class) :in slots
+            :when (and regex (capture-regex-function regex))
+            :do (setf (gethash id (buffer-references (current-buffer)))
+                      (find-ref regex class)))))))
+
+(defgeneric capture-reference (position class))
+
+(defun %get-reference (references)
+  (alexandria:when-let* ((name-references (mapcar #'reference-name references))
+                         (item
+                          (prompt-for-string "Navigate: "
+                                             :completion-function (lambda (x) (completion-strings x name-references))
+
+                                             :test-function (lambda (name)
+                                                              (member name name-references :test #'string=)))))
+    (find item references :key #'reference-name :test #'string=)))
+
+(defgeneric navigate-reference (references))
+
+(defmethod navigate-reference ((type String))
+  (alexandria:when-let ((references (gethash type (buffer-references (current-buffer)))))
+    (%get-reference references)))
+
+(defmethod navigate-reference ((references List))
+  (%get-reference references))
+
+(defgeneric move-to-reference (reference))
+
+(defmethod move-to-reference ((reference reference))
+  (let ((location (reference-point reference)))
+    (move-point (current-point) location)))
+
+(defmethod move-to-reference (reference)
+  (message "Not reference available in current buffer."))
+
+(defun check-change ()
+  (cond 
+   ((null (buffer-references (current-buffer)))
+    (search-references 
+     (variable-value 'lem/language-mode:detective-search :buffer)))
+   
+   ((changed-disk-p (current-buffer))
```

Comment:
It will maybe need to be updated if the buffer has been edited since it was last retrieved.

Reply:
It seems that lem:buffer-modified-tick can be used to detect this.

Reply:
Maybe something like this?
```lisp
(defun check-change ()
  (cond 
   ((null (buffer-references (current-buffer)))
    (search-references 
     (variable-value 'lem/language-mode:detective-search :buffer)))
   
   ((not (eql (buffer-value (current-buffer) 'prev-tick)
              (buffer-modified-tick (current-buffer))))
    
    (setf (buffer-value (current-buffer) 'prev-tick)
          (buffer-modified-tick (current-buffer)))
    (search-references (variable-value 'lem/language-mode:detective-search :buffer)))))

```

Reply:
Yes, it looks good.

Reply:
Added :+1: